### ISSUE TITLE
Excessive runtime for nested `OrderedDict` instances

### ISSUE THREAD

**Issue by Glyphack** [astral-sh/ty#3123]
### Summary

I found this to be slow when type checking:

```
from collections import OrderedDict

OrderedDict(OrderedDict(OrderedDict(OrderedDict(OrderedDict(OrderedDict(("one", 1)))))))
```
[Playground](https://play.ty.dev/dfcc47f0-6184-4a47-8968-ee527b79c9b7)

This takes about 11 seconds on my machine.

The example was found using a fuzzer.
While this example is not a real program someone would write, I'm submitting it in case the root cause can slow down `ty` in other situations that are useful. Feel free to close if it's unrelated.

### Version

0.15.7+22 (dd6d84305 2026-03-21)

**Comment by ibraheemdev** [comment]
It looks like there is exponential blowup here because there are multiple matching overloads for each nested calls, causing us to infer the subsequent nested call multiple times with each parameter type as type context. This is similar to what https://github.com/astral-sh/ruff/pull/22066 attempted to address. We should be able to filter the type context out as ineffective based on the argument type, but that seems tricky to implement consistently. It's possible that we should just bail out of multi-inference for nested generic calls (which would lead to false positives in some, hopefully rare, cases).

**Comment by ibraheemdev** [comment]
Actually, we may be able to get away with having a shared `HashMap<(Expr, TypeContext), Type>` cache across speculated type inference attempts.
